### PR TITLE
pageserver: make TenantsMap shard-aware

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

When using TenantId as the key, we are unable to handle multiple tenant shards attached to the same pageserver for the same tenant ID.  This is an expected scenario if we have e.g. 8 shards and 5 pageservers.

## Summary of changes

- TenantsMap is now a BTreeMap instead of a HashMap: this enables looking up by range.  In future, we will need this for page_service, as incoming requests will just specify the Key, and we'll have to figure out which shard to route it to.
- A new key type TenantShardId is introduced, to act as the key in TenantsMap, and as the id type in external APIs.  Its human readable serialization is backward compatible with TenantId, and also forward-compatible as long as sharding is not actually used (when we construct a TenantShardId with ShardCount(0), it serializes to an old-fashioned TenantId).
- Essential tenant APIs are updated to accept TenantShardIds: tenant/timeline create, tenant delete, and /location_conf.  These are the APIs that will enable driving sharded tenants.  Other apis like /attach /detach /load /ignore will not work with sharding: those will soon be deprecated and replaced with /location_conf as part of the live migration work.

Closes: #5787

### CODE REVIEW

File: libs/pageserver_api/src/shard.rs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+        }
+    }
+}
+
+impl Serialize for TenantShardId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        if serializer.is_human_readable() {
```
Comment (arpad-m) [inline libs/pageserver_api/src/shard.rs:128]: this stuff is pretty cool :+1:

**Review by arpad-m — CHANGES_REQUESTED** [review]
it's pretty cool to distinguish human readable vs non human readable, but the round trip tests only use strings. Could you also add a round trip using a binary format like bincode so that the `From<[u8; 18]>` path is also excercised? That way we make sure that we got the indices right.

File: Cargo.toml [hunk 2]
```diff
@@ -92,6 +92,7 @@ jsonwebtoken = "8"
 libc = "0.2"
 md5 = "0.7.0"
 memoffset = "0.8"
+mur3 = "0.1.0"
```
[the remark this answers is not in the document]
Comment (jcsp) [reply Cargo.toml:95]: Good catch.  No, it's not.  Removed.

File: libs/pageserver_api/src/shard.rs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+            })
+        } else {
+            Err(hex::FromHexError::InvalidStringLength)
+        }
+    }
+}
+
+impl From<[u8; 18]> for TenantShardId {
+    fn from(b: [u8; 18]) -> Self {
+        let mut tenant_id_bytes: [u8; 16] = [0; 16];
```
[the remark this answers is not in the document]
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:112]: Applied by hand.

File: libs/pageserver_api/src/shard.rs (hunk 1 above)
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:128]: Yeah, I can't take credit though :-)  The structure is just a copy paste of what was recently added for TenantId

File: libs/pageserver_api/src/shard.rs [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+        tenant_id_bytes.copy_from_slice(&b[0..16]);
+
+        Self {
+            tenant_id: TenantId::from(tenant_id_bytes),
+            shard_number: ShardNumber(b[16]),
+            shard_count: ShardCount(b[16]),
```
[the remark this answers is not in the document]
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:118]: You're right, fixed.

**Comment by jcsp** [comment]
fafef4a08 fixes the binary encoding (we don't use it anywhere yet, but it's worth keeping for future use) and adds tests for it.

File: libs/pageserver_api/src/shard.rs [hunk 5]
```diff
@@ -18,6 +18,20 @@ impl ShardNumber {
     pub const MAX: Self = Self(u8::MAX);
 }
 
+/// TenantShardId identify the units of work for the Pageserver.
+///
+/// Historically, tenants only had one shard, and were identified
```
Comment (arpad-m) [inline libs/pageserver_api/src/shard.rs:23]: We still support the 1-shard case, right? Maybe it should read "Historically, tenants always had one shard"?

File: libs/pageserver_api/src/shard.rs [hunk 6]
```diff
@@ -216,7 +230,25 @@ mod tests {
...
+        let encoded = bincode::serialize(&example).unwrap();
+        let decoded = bincode::deserialize(&encoded).unwrap();
+
+        assert_eq!(example, decoded);
+
+        Ok(())
+    }
+
+    #[test]
+    fn tenant_shard_id_compat_backward() -> Result<(), hex::FromHexError> {
```
Comment (arpad-m) [inline libs/pageserver_api/src/shard.rs:249]: `tenant_shard_id_backward_compat` reads better I think.

File: libs/pageserver_api/Cargo.toml [hunk 7]
```diff
@@ -17,5 +17,10 @@ postgres_ffi.workspace = true
 enum-map.workspace = true
 strum.workspace = true
 strum_macros.workspace = true
+url.workspace = true
```
Comment (arpad-m) [inline libs/pageserver_api/Cargo.toml:20]: is this one used?

File: pageserver/src/tenant/mgr.rs [hunk 8]
```diff
@@ -87,10 +89,37 @@ pub(crate) enum TenantsMap {
...
+    match (slot_a, slot_b) {
+        (None, None) => None,
+        (Some(slot), None) => {
+            // Exactly one matching slot
+            Some(slot)
+        }
+        (Some(_slot_a), Some(_slot_b)) => {
+            // Multiple shards for this tenant: cannot handle this yet.
+            // TODO(sharding): callers of get() should be shard-aware.
+            todo!()
```
Comment (arpad-m) [inline pageserver/src/tenant/mgr.rs:119]: It would be great if there was a panic message here.

**Review by arpad-m — APPROVED** [review]
I think it's good to go, except for the discussions I've opened. Glad that me asking for tests found another bug in the binary encoding :).

File: libs/pageserver_api/Cargo.toml (hunk 7 above)
Comment (jcsp) [reply libs/pageserver_api/Cargo.toml:20]: Will be in subsequent changes to shard.rs, but not now.  Removed.

File: libs/pageserver_api/src/shard.rs (hunk 6 above)
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:249]: I'm ambivalent, but it costs nothing to change it, so changed :-)

File: libs/pageserver_api/src/shard.rs (hunk 5 above)
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:23]: I've fleshed this comment out a bit.

File: pageserver/src/tenant/mgr.rs (hunk 8 above)
Comment (jcsp) [reply pageserver/src/tenant/mgr.rs]: added

File: libs/pageserver_api/src/shard.rs (hunk 1 above)
Comment (koivunej) [reply libs/pageserver_api/src/shard.rs:147]: related to the huge work which went in and from where this was copypasted: It'd be great if we had just one bincode roundtrip test for this, as I assume these are going to be important.

File: libs/pageserver_api/src/shard.rs [hunk 9]
```diff
@@ -0,0 +1,297 @@
...
+        // that the same behavior applies when we have used `unsharded()` to
+        // construct a TenantShardId.
+        let example = TenantShardId::unsharded(TenantId::generate());
+        let encoded = bincode::serialize(&example).unwrap();
+        let decoded = bincode::deserialize::<TenantShardId>(&encoded).unwrap();
+        assert_eq!(example, decoded);
+
+        Ok(())
+    }
+}
```
Comment (koivunej) [inline libs/pageserver_api/src/shard.rs:297]: Please make the tests deterministic (or copy one from any test which has a non-zero tenant_id) and add expected serlialized bytes here. You can use `utils::Hex` for initial:

```
let expected: [u8; 0] = [];

assert_eq!(Hex(&encoded), Hex(&expected));
```

And then just paste the expected.

**Review by koivunej — CHANGES_REQUESTED** [review]
Expected bytes in serialization tests; otherwise it's difficult to get the added serialization tests fail.

File: libs/pageserver_api/src/shard.rs [hunk 10]
```diff
@@ -0,0 +1,297 @@
...
+        };
+
+        let encoded = format!("{example}");
+
+        let decoded = TenantShardId::from_str(&encoded)?;
+
+        assert_eq!(example, decoded);
+
+        Ok(())
+    }
```
Comment (koivunej) [inline libs/pageserver_api/src/shard.rs:236]: Similarly for these string tests it would be great to have the expected string.

**Comment by arpad-m** [comment]
> otherwise it's difficult to get the added serialization tests fail.

Not saying that a binary test isn't helpful but note that the serialization test has caught already one bug in the (de)serialization code that John then subsequently fixed. another bug was caught by me looking at the code, that's why I suggested adding tests for the binary format in the first place.

File: libs/pageserver_api/src/shard.rs (hunk 9 above)
Comment (jcsp) [reply libs/pageserver_api/src/shard.rs:321]: yep, that's worth doing: added f18930c20e6e22af6af2a6ef0c7dd78d87962d5d

**Review by koivunej — APPROVED** [review]

**Pull request merged**